source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -2452,7 +2460,234 @@ export class MediaSource extends SerializableObject {
+        super(matches);
+
+        if (matches.length === 2) {
+            this._videoId = matches[1];
+        }
+    }
+
+    abstract getEmbedVideoUrl(): string;
+
+    canPlay(): boolean {
+        return this._videoId !== undefined;
+    }
+
+    render(): HTMLElement {
+        let container = document.createElement("div");
+        container.style.position = "relative";
+        container.style.width = "100%";
+        container.style.height = "0";
+        container.style.paddingBottom = "56.25%";
+
+        let iFrame = document.createElement("iframe");
+        iFrame.style.position = "absolute";
+        iFrame.style.top = "0";
+        iFrame.style.left = "0";
+        iFrame.style.width = "100%";
+        iFrame.style.height = "100%";
+        iFrame.src = this.getEmbedVideoUrl();
+        iFrame.frameBorder = "0";
+        iFrame.allow = "accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture";
+        iFrame.allowFullscreen = true;
+        
+        container.appendChild(iFrame);
+
+        return container;
+    }
+
+    get videoId(): string | undefined {
+        return this._videoId;
+    }
+}
+
+export class VimeoPlayer extends IFrameMediaMediaPlayer {
+    async fetchVideoDetails(): Promise<void> {
+        const oEmbedUrl = `https://vimeo.com/api/oembed.json?url=${this.getEmbedVideoUrl()}`;
+
+        let response = await fetch(oEmbedUrl);
+
+        if (response.ok) {
+            let json = await response.json();
+
+            this.posterUrl = json["thumbnail_url"];
+        }
+    }
+
+    getEmbedVideoUrl(): string {
+        return `https://player.vimeo.com/video/${this.videoId}?autoplay=1`;
+    }
+}
+
+export class DailymotionPlayer extends IFrameMediaMediaPlayer {
+    async fetchVideoDetails(): Promise<void> {
+        const apiUrl = `https://api.dailymotion.com/video/${this.videoId}?fields=thumbnail_720_url`;
+
+        let response = await fetch(apiUrl);
+
+        if (response.ok) {
+            let json = await response.json();
+
+            this.posterUrl = json["thumbnail_720_url"];
+        }
+    }
+
+    getEmbedVideoUrl(): string {
+        return  `https://www.dailymotion.com/embed/video/${this.videoId}?autoplay=1`;
+    }
+}
+
+export class YouTubePlayer extends IFrameMediaMediaPlayer {
+    async fetchVideoDetails(): Promise<void> {
+        this.posterUrl = this.videoId ? `https://img.youtube.com/vi/${this.videoId}/maxresdefault.jpg` : undefined;
+    }
+
+    getEmbedVideoUrl(): string {
+        return `https://www.youtube.com/embed/${this.videoId}?autoplay=1`;
+    }
+}
+
+export interface ICustomMediaPlayer {
+    urlPatterns: RegExp[];
+    createMediaPlayer: (matches: RegExpExecArray) => CustomMediaPlayer
+}
+
 export class Media extends CardElement {
+    static customMediaPlayers: ICustomMediaPlayer[] = [
+        {
+            urlPatterns: [ /https?:\/\/www.youtube.com\/watch\?v=([\w\d-_]+)&?/ig, /https?:\/\/youtu.be\/([\w\d-_]+)/ig ],
```

Comment:
this regex will break if e.g. the youtube url includes a starting time index:

`https://www.youtube.com/watch?v=dQw4w9WgXcQ&t=66s`

Reply:
The first is ready for that, but not the second. Good catch.

Reply:
Updated the regular expressions for all three providers.
Added start time support for YouTube.